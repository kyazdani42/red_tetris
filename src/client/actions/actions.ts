import { Action, ActionFunctionAny, createAction } from 'redux-actions';

import { AppState } from '../reducers/app';

import {
  CREATE_ROOM,
  HANDLE_KEY_PRESS,
  JOIN_ROOM,
  LEAVE_ROOM,
  SET_GAME_DATA,
  SET_KEY,
  SET_OPTIONS,
  SET_PLAYER_NAME,
  SET_ROOMS,
  SET_SOCKET,
} from '../constants';

export const setRooms: ActionFunctionAny<Action<RoomType[]>> = createAction(SET_ROOMS);

export const createRoom: ActionFunctionAny<Action<string>> = createAction(CREATE_ROOM);
export const joinRoom: ActionFunctionAny<Action<undefined>> = createAction(JOIN_ROOM);
export const leaveRoom: ActionFunctionAny<Action<undefined>> = createAction(LEAVE_ROOM);

export const setSocket: ActionFunctionAny<Action<SocketIOClient.Socket>> = createAction(SET_SOCKET);

export const setGameData: ActionFunctionAny<Action<GameProps>> = createAction(SET_GAME_DATA);
export const setKey: ActionFunctionAny<Action<keyType>> = createAction(SET_KEY);
export const handleKeyPress: ActionFunctionAny<Action<keyType>> = createAction(HANDLE_KEY_PRESS);

export const setPlayerName: ActionFunctionAny<Action<string>> = createAction(SET_PLAYER_NAME);
export const setOptions: ActionFunctionAny<Action<AppState['options']>> = createAction(SET_OPTIONS);
