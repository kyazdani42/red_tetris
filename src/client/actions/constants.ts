export const SET_ROOMS = 'SET_ROOMS';
export const JOIN_ROOM = 'HANDLE_JOIN_ROOM';

export const CREATE_ROOM = 'CREATE_ROOM';
export const LEAVE_ROOM = 'LEAVE_ROOM';

export const SET_SOCKET = 'SET_SOCKET';

export const SET_GAME_DATA = 'SET_GAME_DATA';
