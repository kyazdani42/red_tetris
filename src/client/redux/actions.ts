import { Action, ActionFunctionAny, createAction } from 'redux-actions';

import { AppState } from './reducer';

import {
  CREATE_ROOM,
  GET_SCORES,
  HANDLE_KEY_PRESS,
  JOIN_ROOM,
  LEAVE_ROOM,
  SET_ERROR,
  SET_GAME_DATA,
  SET_KEY,
  SET_MODAL,
  SET_MUSIC_PLAYING,
  SET_OPTIONS,
  SET_PLAYER_NAME,
  SET_ROOMS,
  SET_SOCKET,
  SET_TOKEN,
} from './constants';

export const setRooms: ActionFunctionAny<Action<RoomType[]>> = createAction(SET_ROOMS);

export const createRoom: ActionFunctionAny<Action<string>> = createAction(CREATE_ROOM);
export const joinRoom: ActionFunctionAny<Action<undefined>> = createAction(JOIN_ROOM);
export const leaveRoom: ActionFunctionAny<Action<undefined>> = createAction(LEAVE_ROOM);
export const setModal: ActionFunctionAny<Action<boolean>> = createAction(SET_MODAL);

export const setSocket: ActionFunctionAny<Action<SocketIOClient.Socket>> = createAction(SET_SOCKET);

export const setGameData: ActionFunctionAny<Action<GameProps>> = createAction(SET_GAME_DATA);
export const setKey: ActionFunctionAny<Action<keyType>> = createAction(SET_KEY);
export const handleKeyPress: ActionFunctionAny<Action<keyType>> = createAction(HANDLE_KEY_PRESS);

export const setPlayerName: ActionFunctionAny<Action<string>> = createAction(SET_PLAYER_NAME);
export const setError: ActionFunctionAny<Action<string | null>> = createAction(SET_ERROR);
export const setOptions: ActionFunctionAny<Action<AppState['options']>> = createAction(SET_OPTIONS);
export const setToken: ActionFunctionAny<Action<string>> = createAction(SET_TOKEN);
export const setMusicPlaying: ActionFunctionAny<Action<undefined>> = createAction(SET_MUSIC_PLAYING);

export const getScores: ActionFunctionAny<Action<undefined>> = createAction(GET_SCORES);
