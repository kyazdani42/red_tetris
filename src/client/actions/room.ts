import { Action, ActionFunctionAny, createAction } from 'redux-actions';
import { RoomType } from '../types';
import { HANDLE_JOIN_ROOM, SET_ROOMS, SET_SELECTED_ROOM } from './constants';

export const setRooms: ActionFunctionAny<Action<RoomType[]>> = createAction(SET_ROOMS);
export const setSelectedRoom: ActionFunctionAny<Action<RoomType>> = createAction(SET_SELECTED_ROOM);
export const handleJoinRoom: ActionFunctionAny<Action<undefined>> = createAction(HANDLE_JOIN_ROOM);
