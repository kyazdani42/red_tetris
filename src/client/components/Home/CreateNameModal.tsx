import * as React from 'react';
import { connect } from 'react-redux';

import { State } from '../../redux/store';
import {
  CreateRoomStyle,
  ErrorStyle,
  InputWrapper,
  LabelStyle,
  ModalStyle,
  ModalWrapper,
  NameInputStyle
} from './styles';

interface Props {
  playerName: string | null;
  handleDispatch: (username: string) => void;
  setDisplayModal: (d: boolean) => void;
}

export const CreateNameModal: React.SFC<Props> = ({ handleDispatch, setDisplayModal, playerName }) => {
  const [error, setError] = React.useState(null);
  if (playerName) {
    handleDispatch(playerName);
    return null;
  }
  return (
    <ModalWrapper onClick={handleRemoveModal(setDisplayModal)} id="modal-name">
      <ModalStyle>
        <InputWrapper>
          <LabelStyle>Pick a username</LabelStyle>
          <NameInputStyle onKeyDown={handleKeyDown(handleDispatch, setError)} autoFocus={true} />
          {error ? <ErrorStyle>{error}</ErrorStyle> : null}
        </InputWrapper>
        <CreateRoomStyle onClick={handleCreateRoom(handleDispatch, setError)}>
          Launch Game
        </CreateRoomStyle>
      </ModalStyle>
    </ModalWrapper>
  );
};

const handleRemoveModal = (setDisplayModal: any) => (e: any) => {
  if (e.target.id === 'modal-name') {
    setDisplayModal(false);
  }
};

const handleCreateRoom = (dispatch: (username: string) => void, setError: any) => (e: any) => {
  const value = e.target.previousSibling.firstChild.nextElementSibling.value;
  if (!value.length) {
    setError('please enter something');
  } else {
    dispatch(value);
  }
};

const handleKeyDown = (dispatch: (username: string) => void, setError: any) => (e: any) => {
  setError(null);
  if (e.key === 'Enter') {
    if (!e.target.value.length) {
      setError('please enter something');
    } else {
      dispatch(e.target.value);
    }
  }
};

const mapStateToProps = (state: State, ownProps: any) => ({
  playerName: state.app.playerName,
  ...ownProps
});

export default connect(mapStateToProps)(CreateNameModal);
