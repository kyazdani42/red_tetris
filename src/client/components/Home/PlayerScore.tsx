import * as React from 'react';
import { connect } from 'react-redux';

import { State } from '../../redux/store';
import { PlayerScoreDisplay, PlayerScoreWrapper } from './styles';

interface Props {
  playerScore: BestScore | null;
}

export const PlayerScore: React.SFC<Props> = ({ playerScore }) =>
  (playerScore && (
    <PlayerScoreWrapper>
      <PlayerScoreDisplay>Hello {playerScore.name}</PlayerScoreDisplay>
      <PlayerScoreDisplay>Your best Score: {playerScore.bestScore}</PlayerScoreDisplay>
      <PlayerScoreDisplay>
        You played {playerScore.gamesPlay} games and won {playerScore.multiPlayersWin} times
      </PlayerScoreDisplay>
    </PlayerScoreWrapper>
  )) ||
  null;

export const mapStateToProps = (state: State) => ({
  playerScore: state.app.playerScore
});

export default connect(mapStateToProps)(PlayerScore);
