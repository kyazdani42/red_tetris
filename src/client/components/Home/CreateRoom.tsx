import * as React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { createRoom } from '../../actions/actions';
import { CreateNameModal } from './CreateNameModal';

interface Props {
  dispatchCreateRoom: (username: string) => void;
}

const Button = styled.div`
  font-family: ${props => props.theme.fonts.title};
  background-color: ${props => props.theme.colors.primary.dark};
  border-radius: 4px;
  color: #fff;
  width: 35%;
  padding: 5px;
  font-size: 22px;
  border: 1px solid rgba(0, 0, 0, 0.5);
  text-align: center;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.2), 0 3px 6px rgba(0, 0, 0, 0.3);
  user-select: none;
  cursor: pointer;
  :hover {
    opacity: 0.94;
  }
  @media screen and (max-width: 500px) {
    font-size: 18px;
  }
`;

const Wrapper = styled.div`
  margin-top: 12px;
  display: flex;
  justify-content: center;
  align-items: flex-end;
`;

export const CreateRoomButton: React.SFC<Props> = ({ dispatchCreateRoom }) => {
  const [displayModal, setDisplayModal] = React.useState(false);
  let modal;
  if (displayModal) {
    modal = <CreateNameModal setDisplayModal={setDisplayModal} dispatch={dispatchCreateRoom} />;
  } else {
    modal = null;
  }
  return (
    <React.Fragment>
    <Wrapper>
      <Button onClick={handleClick(setDisplayModal)}>
        New Game
      </Button>
    </Wrapper>
    {modal}
    </React.Fragment>
  );
};

const handleClick = (setDisplayModal: any) => (e: any) => {
  const target = e.currentTarget;
  target.style.opacity = '0.90';
  setTimeout(() => { target.style.opacity = '1'; }, 100);
  setDisplayModal(true);
};

const mapDispatchToProps = (dispatch: any) => ({
  dispatchCreateRoom: (username: string) => dispatch(createRoom(username))
});

export default connect(undefined, mapDispatchToProps)(CreateRoomButton);
