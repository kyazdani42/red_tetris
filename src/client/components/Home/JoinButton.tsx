import * as React from 'react';
import { connect } from 'react-redux';
import { withTheme } from 'styled-components';

import { joinRoom, setModal } from '../../redux/actions';
import { State } from '../../redux/store';
import CreateNameModal from './CreateNameModal';
import { JoinButtonStyle } from './styles';
import { handleClick } from './utils';

interface Props {
  dispatchJoinRoom: (id: string) => (name: string) => void;
  dispatchSetModal: (display: boolean) => void;
  displayModal: boolean;
  roomId: string;
}
export const JoinButton: React.SFC<Props> = ({
  dispatchJoinRoom,
  roomId,
  dispatchSetModal,
  displayModal
}) => {
  let modal;
  if (displayModal) {
    modal = (
      <CreateNameModal
        className="name-modal"
        setDisplayModal={dispatchSetModal}
        handleDispatch={dispatchJoinRoom(roomId)}
      />
    );
  } else {
    modal = null;
  }
  return (
    <React.Fragment>
      <JoinButtonStyle onClick={handleClick(dispatchSetModal)}>
        <span>Join</span>
        <Svg />
      </JoinButtonStyle>
      {modal}
    </React.Fragment>
  );
};

const Svg = () => (
  <svg width="16" height="16" viewBox="0 0 24 24">
    <path d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z" />
    <path fill="none" d="M0 0h24v24H0z" />
  </svg>
);

const mapStateToProps = (state: State, { roomId }: { roomId: string }) => ({
  roomId,
  displayModal: state.app.modal
});

const mapDispatchToProps = (dispatch: any) => ({
  dispatchJoinRoom: (id: string) => (name: string) =>
    dispatch(joinRoom({ room: id, playerName: name })),
  dispatchSetModal: (display: boolean) => dispatch(setModal(display))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTheme(JoinButton as any));
