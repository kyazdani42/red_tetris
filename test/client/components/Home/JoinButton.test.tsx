// tslint:disable
import * as React from 'react';
import { shallow } from 'enzyme';

import { JoinButton, mapStateToProps, mapDispatchToProps } from '../../../../src/client/components/Home/JoinButton';
import { setModal, joinRoom } from '../../../../src/client/redux/actions';

describe('testing the component', () => {
  const props = {
    dispatchJoinRoom: () => () => null,
    dispatchSetModal: () => null,
    displayModal: true,
    roomId: 'test-id'
  };
  const wrapper = shallow(<JoinButton {...props} />);
  it('smoke tests the component', () => {
    expect(wrapper.update()).toMatchSnapshot();
  });
  it('checks that join button exists', () => {
    expect(wrapper.find('.name-modal').exists()).toBeTruthy();
  })
  it('checks that join button doesn\'t exists', () => {
    wrapper.setProps({ displayModal: false });
    expect(wrapper.find('.name-modal').exists()).toBeFalsy();
  })
});

describe('testing mapStateToProps', () => {
  const state: any = { app: { modal: 'modal' } };
  const ownProps = { roomId: 'roomIdTest' };
  it('test the mapping', () => {
    const expected = { displayModal: 'modal', roomId: 'roomIdTest' };
    expect(mapStateToProps(state, ownProps)).toEqual(expected);
  })
})

describe('testing mapDispatchToProps', () => {
  const dispatch = jest.fn();
  const mapper = mapDispatchToProps(dispatch);
  it('checks the keys', () => {
    expect(Object.keys(mapper)).toEqual(['dispatchJoinRoom', 'dispatchSetModal']);
  })
  it('checks the dispatch functions', () => {
    mapper.dispatchJoinRoom('test')('innerTest');
    expect(dispatch).toHaveBeenCalledWith(joinRoom({ room: 'test', playerName: 'innerTest' }));
    mapper.dispatchSetModal(true);
    expect(dispatch).toHaveBeenCalledWith(setModal(true));
  })
})