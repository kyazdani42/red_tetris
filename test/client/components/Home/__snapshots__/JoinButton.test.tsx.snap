// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the component smoke tests the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    dispatchJoinRoom={[Function]}
    dispatchSetModal={[Function]}
    displayModal={true}
    roomId="test-id"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <UNDEFINED
          onClick={[Function]}
        >
          <span>
            Join
          </span>
          <Svg />
        </UNDEFINED>,
        <Connect(Component)
          className="name-modal"
          handleDispatch={[Function]}
          setDisplayModal={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <span>
              Join
            </span>,
            <Svg />,
          ],
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Join",
            },
            "ref": null,
            "rendered": "Join",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htoDjs",
            "isStatic": false,
            "rules": Array [
              "
  height: 30px;
  width: 80px;
  background-color: ",
              [Function],
              ";
  border: 1px solid rgba(0, 0, 0, 0.2);
  font-size: 17px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.05), 0 3px 6px rgba(0, 0, 0, 0.13);
  padding-top: 1px;
  padding-left: 5px;
  margin-left: 3px;
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;
  align-content: center;
  transition: box-shadow 300ms;
  user-select: none;
  cursor: pointer;
  &:hover {
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.2), 0 3px 6px rgba(0, 0, 0, 0.3);
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htoDjs",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "name-modal",
          "handleDispatch": [Function],
          "setDisplayModal": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <UNDEFINED
            onClick={[Function]}
          >
            <span>
              Join
            </span>
            <Svg />
          </UNDEFINED>,
          <Connect(Component)
            className="name-modal"
            handleDispatch={[Function]}
            setDisplayModal={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <span>
                Join
              </span>,
              <Svg />,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Join",
              },
              "ref": null,
              "rendered": "Join",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htoDjs",
              "isStatic": false,
              "rules": Array [
                "
  height: 30px;
  width: 80px;
  background-color: ",
                [Function],
                ";
  border: 1px solid rgba(0, 0, 0, 0.2);
  font-size: 17px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.05), 0 3px 6px rgba(0, 0, 0, 0.13);
  padding-top: 1px;
  padding-left: 5px;
  margin-left: 3px;
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;
  align-content: center;
  transition: box-shadow 300ms;
  user-select: none;
  cursor: pointer;
  &:hover {
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.2), 0 3px 6px rgba(0, 0, 0, 0.3);
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htoDjs",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "name-modal",
            "handleDispatch": [Function],
            "setDisplayModal": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
